internal/controllers/sacd_helpers.go

```diff
@@ -0,0 +1,115 @@
+package controllers
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/DIMO-Network/cloudevent"
+	"github.com/DIMO-Network/shared"
+	"github.com/DIMO-Network/token-exchange-api/internal/models"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+func userGrantMap(record *models.PermissionRecord, nftAddr string, tokenID int64) (map[string]bool, map[string]map[string]*shared.StringSet, error) {
+	userPermGrants := make(map[string]bool)
+	cloudEvtGrants := make(map[string]map[string]*shared.StringSet)
+
+	// Aggregates all the permission and attestation grants the user has.
+	for _, agreement := range record.Data.Agreements {
+		now := time.Now()
+		if agreement.EffectiveAt != nil && agreement.EffectiveAt.After(now) {
+			continue
+		}
+
+		if agreement.ExpiresAt != nil && agreement.ExpiresAt.Before(now) {
+			continue
+		}
+
+		switch agreement.Type {
+		case "cloudevent":
+			if err := validAssetDID(agreement.Asset, nftAddr, tokenID); err != nil {
+				return nil, nil, fmt.Errorf("failed to validate attestation asset: %s", agreement.Asset)
+			}
+
+			if agreement.EffectiveAt != nil && !agreement.EffectiveAt.IsZero() && now.Before(*agreement.EffectiveAt) {
+				continue // agreement not yet in effective
+			}
+
+			if agreement.ExpiresAt != nil && !agreement.ExpiresAt.IsZero() && agreement.ExpiresAt.Before(now) {
+				return nil, nil, errors.New("agreement expired")
+			}
+
+			if _, ok := cloudEvtGrants[agreement.EventType]; !ok {
+				cloudEvtGrants[agreement.EventType] = map[string]*shared.StringSet{}
+			}
+
+			source := strings.ToLower(agreement.Source)
```

Comment:
I don't think we want a case insensitive check. We want to be using a real address
```suggestion
			source := strings.ToLower(agreement.Source)
```